### ISSUE TITLE
CustomConverters with enabled method

### ISSUE THREAD

**Issue by lordofthejars** [eclipse/microprofile-config#466]
I think that `Converter` interface should implement a `default` method that allows you to enable a converter. Let me explain with an example:

Currently, the spec offers one converter from String to double and that's fine but suppose I want to implement the next converter: A converter from `percentage` to `double` ie:

```
loan.percentage=10%
```

Then I want to do:

```
@ConfigSource(name="...")
double percentage;
```

And the converter:

```
return Double.parseDouble(element.substring(0, element.indexOf("%"))/100)
```

The problem is that if I set a high priority then for every String to Double this converter will be applied which means that I will need to do something like:

```
if (element.endsWith("%") then ... else Double.parseDouble(element);
```

This is fine but another approach that I think will make converters cleaner will be to have an enable method so I can do:

```
public boolean isEnabled(String element) {
  return element.endsWith("%");
}
```

Then for every converter of type `Double, String` you order by priority and you try to execute the first one calling `isEnabled` if it is true then it is executed the conversion, if not it is ignored and tries the next one (a chain call)

To maintain break-back compatibility the `isEnabled` method could be a default method in the interface returning `true`.

**Comment by jmesnil** [comment]
The solution for your use case would be to create a `Percentage` interface:

```
interface Percentage {
  double value()
}
```

Then you can create a *specific* converter of type `Converter<Percentage>` that would make sure that the string value represents a percentage. It could accept:
* `10%`
* `0.1`
* `.1`
or whatever you decide is a valid textual representation of a percentage.

Your code would simply be:

```
@ConfigSource(name="ratio", default="10%")
Percentage ratio;

...

ratio.value() // -> return the double value of the percentage
```

**Comment by lordofthejars** [comment]
Yes I know I could do this, but then I need to create two new classes for just a simple conversion. I am just thinking on the fast developing of simple things.

**Comment by jmesnil** [comment]
Converters are based on type.

 If you want a `double`, you must use a textual representation that is a valid input for a double. `10%` is not.

Providing a `Percentage` type is correct, self-documenting, type-safe and does not impact other properties that requires proper `double` values.

**Comment by dmlloyd** [comment]
Philosophically I agree with Jeff that different value formats should be represented by different types, however pragmatically it probably isn't *really* a terrible idea to *extend* the set of valid values for a type.  I would shy away from customizing converters that *reduce* the set of valid values or which change the interpretation of values which are normally valid for that type.

To give an example I was considering a change to our implementation to allow Java-style hexadecimal, octal, and binary values to be specified for integral types.  The main problem that prevented me from going through with the change was that octal values, which are of the format `0[0-7]+`, would change the specification-driven interpretation of decimal values with leading zeros.  Using octal values for UNIX-style file permissions was one of my key use cases, so it was a bit of a disappointment.  However, there's nothing really stopping me from still developing such a converter and only using it for certain properties whose syntax is documented, so I may yet at least write the code and use it judiciously rather than installing it as a default.

Adding percentage `%` or even e.g. `‰` and `‱` would expand the range of allowable values without impacting the existing set, and would therefore not be completely unreasonable (even if it was philosophically debatable).

I've often wondered if it might be a good idea to do a little introspection on the converter when instantiating it and being a bit smarter about constructor matching.  For example, for a class implementing `Converter<Foo>`, a parameter named `Converter<Foo>` could possibly be initialized to the next-lower-priority converter for that type.

This might also theoretically allow for generic converters.  If my class implements `Converter<Foo<T>>` and it has a parameter of type `Converter<T>`, then that parameter could be recursively populated by the converter of the delegating type.  This would be another good step towards allowing things like `Optional`, collection, and array conversion to be self-consistent with the rest of the API.
